File: examples/sem/1_sync.cpp
// Copyright (c) Ivan Janevski
#include <thread.h>
#include <sem.h>

Sem sa = 1, sb = 0;

void a() {
    while (true) {
        wait(sa);
        print(RED "a");
		signal(sb);
    }
}

void b() {
    while (true) {
        wait(sb);
        print(GREEN "b");
		signal(sa);
    }
}


int main() {
    Thread ta = createThread(a);
    Thread tb = createThread(b);
    join(ta);
    join(tb);
    return 0;
}

File: examples/golang/1_sync.go
/**
 * (c) Ivan Janevski
 */
package main

import ( "fmt" )

const RED = "\x1b[31m"
const GREEN = "\x1b[32m"
const NOCOLOR = "\x1b[0m"

// channels
var aChannel, bChannel chan string
var aFinished, bFinished chan bool

// routines
func a() {
    for {
        msg := <- aChannel
        fmt.Print(GREEN + msg + NOCOLOR)
        bChannel <- "b"
    }
    aFinished <- true
}

func b() {
    for {
        msg := <- bChannel
        fmt.Print(RED + msg + NOCOLOR)
        aChannel <- "a"
    }
    bFinished <- true
}

// main
func main() {
    // initialize channels
    aChannel = make(chan string, 1)
    bChannel = make(chan string, 1)
    aFinished = make(chan bool)
    bFinished = make(chan bool)
    aChannel <- "a"

    // start go routines
    go a()
    go b()

    // wait for go routines
    <- aFinished
    <- bFinished
}

